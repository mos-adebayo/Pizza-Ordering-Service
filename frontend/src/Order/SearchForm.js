import React from 'react';
import PropTypes from 'prop-types';
import { Formik, Field, ErrorMessage } from 'formik';
import { Form, Button, Row, Col, Spinner } from  "react-bootstrap";
import * as Yup from "yup";

const validationSchema = Yup.object().shape({
    customerName: Yup.string().required('Please enter customer name'),
    address: Yup.string().required('Please enter your address'),
    quantity: Yup.number()
        .required("Please enter quantity")
        .positive("Please enter valid quantity")
        .integer("Please enter valid quantity"),
    pizzaSizeId: Yup.string().required('Please select size'),
    pizzaTypeId: Yup.string().required('Please select type'),
});

const initialValues = { customerName: "", quantity: "", pizzaSizeId: "", pizzaTypeId: "", address: "" };

const SearchForm = ({ pizzaSizes, pizzaTypes, formHandler, requesting }) => {
    return (
        <Formik
            initialValues={initialValues}
            validationSchema={validationSchema}
            onSubmit={(values, ...props ) => {
                formHandler(values, props[0].resetForm);
            }}
        >
            {({ handleSubmit }) => (
                <form onSubmit={handleSubmit}>
                    <Row>
                        <Col>
                            <Form.Group>
                                <Form.Label> Customer </Form.Label>
                                <Field type="text" name="customerName" placeholder={"Enter customer name"} className={'form-control'} />
                                <ErrorMessage name="customerName" component="div" className={"text-danger"} />
                            </Form.Group>
                        </Col>
                        <Col>
                            <Form.Group>
                                <Form.Label> Quantity </Form.Label>
                                <Field type="text" name="quantity" placeholder="Enter quantity" className={'form-control'} />
                                <ErrorMessage name="quantity" component="div" className={"text-danger"}/>
                            </Form.Group>
                        </Col>
                    </Row>
                    <Row>
                        <Col>
                            <Form.Group>
                                <Form.Label> Pizza Size </Form.Label>
                                <Field component="select" name="pizzaSizeId" className={'form-control'} >
                                    <option value="-1">Select size</option>
                                    {
                                        pizzaSizes.map((item, index) => <option value={item.id} key={index}>
                                            {`${item.name}` }
                                        </option>)
                                    }
                                </Field>
                                <ErrorMessage name="pizzaSizeId" component="div"  className={"text-danger"}/>
                            </Form.Group>
                        </Col>
                        <Col>
                            <Form.Group>
                                <Form.Label> Pizza Type </Form.Label>
                                <Field component="select" name="pizzaTypeId" className={'form-control'} >
                                    <option value="-1">Select type</option>
                                    {
                                        pizzaTypes.map((item, index) => <option value={item.id} key={index}>
                                            {`${item.name}` }
                                        </option>)
                                    }
                                </Field>
                                <ErrorMessage name="pizzaTypeId" component="div" className={"text-danger"} />
                            </Form.Group>
                        </Col>
                    </Row>
                    <Row>
                        <Col>
                            <Form.Group>
                                <Form.Label> Address </Form.Label>
                                <Field type="text" name="address" placeholder={"Enter address"} className={'form-control'} />
                                <ErrorMessage name="address" component="div" className={"text-danger"} />
                            </Form.Group>
                        </Col>
                    </Row>
                    <Row>
                        <Col>
                            <Form.Group>
                                {
                                    requesting ? (
                                        <Button disabled>
                                            <Spinner animation="grow" />
                                        </Button>
                                    ) : (
                                        <Button type="submit" className={'submit-btn'}>
                                            Make Order
                                        </Button>
                                    )
                                }
                            </Form.Group>
                        </Col>
                    </Row>
                </form>
            )}
        </Formik>
    );
};

SearchForm.propTypes = {
    pizzaSizes: PropTypes.array.isRequired,
    pizzaTypes: PropTypes.array.isRequired,
    formHandler: PropTypes.func.isRequired
};

export default SearchForm;