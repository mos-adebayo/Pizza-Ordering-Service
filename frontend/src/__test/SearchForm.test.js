import React from "react";
import { configure, mount } from "enzyme";
import ReactDOM from "react-dom";
import renderer from 'react-test-renderer';
import Adapter from 'enzyme-adapter-react-16';

import SearchForm  from "../Order/SearchForm";

//Enzyme Configurations
configure({ adapter: new Adapter() });

function setup() {
    const props = {
        pizzaSizes: [],
        pizzaTypes: [],
        requesting: false,
        formHandler: jest.fn()
    };

    const enzymeWrapper = mount(
        <SearchForm
            {...props}
        />
    );
    return {
        props,
        enzymeWrapper
    }
}

test('Search Form renders without crashing', () => {
    const  { props } = setup();

    const div = document.createElement('div');
    ReactDOM.render(
       <SearchForm {...props} />,
        div);
    ReactDOM.unmountComponentAtNode(div);
});

test('Search Form components snapshot', () => {
    const  { enzymeWrapper, props } = setup();

    const component = renderer.create(
        <SearchForm {...props} />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

});

test('should call remove form handler callback when the submit button is clicked ', () => {
    const { enzymeWrapper, props } = setup();

    const target = enzymeWrapper.find('.submit-btn');
    // console.log(target)
    // target.simulate('click');
    // expect(props.removeMarker.mock.calls.length).toBe(1)
    // expect(target.length).toBe(0)
});
