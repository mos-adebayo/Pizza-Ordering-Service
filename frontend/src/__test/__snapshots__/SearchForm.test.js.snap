// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Form components snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <div
        className="form-group"
      >
        <label
          className="form-label"
        >
           Customer 
        </label>
        <input
          className="form-control"
          name="customerName"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="col"
    >
      <div
        className="form-group"
      >
        <label
          className="form-label"
        >
           Quantity 
        </label>
        <input
          className="form-control"
          name="quantity"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <div
        className="form-group"
      >
        <label
          className="form-label"
        >
           Pizza Size 
        </label>
        <select
          className="form-control"
          name="pizzaSizeId"
          onBlur={[Function]}
          onChange={[Function]}
          value=""
        >
          <option
            value="-1"
          >
            Select size
          </option>
        </select>
      </div>
    </div>
    <div
      className="col"
    >
      <div
        className="form-group"
      >
        <label
          className="form-label"
        >
           Pizza Type 
        </label>
        <select
          className="form-control"
          name="pizzaTypeId"
          onBlur={[Function]}
          onChange={[Function]}
          value=""
        >
          <option
            value="-1"
          >
            Select type
          </option>
        </select>
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <div
        className="form-group"
      >
        <label
          className="form-label"
        >
           Address 
        </label>
        <input
          className="form-control"
          name="address"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <div
        className="form-group"
      >
        <button
          className="submit-btn btn btn-primary"
          disabled={false}
          type="submit"
        >
          Make Order
        </button>
      </div>
    </div>
  </div>
</form>
`;
